### PR TITLE
Add new function for encode module

### PR DESCRIPTION
**Pull request by lalalademaxiya1**
Improved encoder

## Description

- Improve the encoder and make it perform better, specifcally add a function called "encoder_insert_state", and will be called in pal set line event callback function when we turn the knob every time.

## Types of Changes

- [x] Core
- [ ] Bugfix
- [x] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by zvecr** [comment]
Core changes should be separate to keyboard changes.

**Comment by lokher** [comment]
> Core changes should be separate to keyboard changes.

We added a new function to encoder.h and encoder.c, which is only invoked by Q2, and it won't break anything, why do we need to separate it? It's kinda wierd if we just add a single function, maybe someone would ask how to use it to fix this issue.

**Comment by zvecr** [comment]
Because that is the process we ask people to follow, as it allows us manage the project easier.

By default core changes also have to go via develop, so i have updated the merge target (this is within the PR checklist).

**Comment by FWest98** [comment]
This change no longer works properly - it causes the rotary encoder to stop functioning altogether. This is due to the new if-condition at https://github.com/qmk/qmk_firmware/blob/master/quantum/encoder.c#L197, introduced in #16812. It prevents the encoder from triggering an action when this code path does not detect a state change - this will almost always be the case since `encoder_insert_state` already updates this state. The best solution I found to this issue is to introduce an additional condition that will trigger the update function to be called after the external update. See: https://github.com/FWest98/OS-QMK_Firmware/commit/95fbce9628191522bf9acb07808812b5c5463152#diff-c8e0671af976b0c896b651211e9605562985027873e5ef0d8a51a74c17aa8a6e

**Comment by spacecakes** [comment]
Any updates on this? Is it still relevant? @lokher

File: quantum/encoder.c [hunk 1]
```diff
@@ -177,3 +177,9 @@ void encoder_update_raw(uint8_t* slave_state) {
     if (changed) last_encoder_activity_trigger();
 }
 #endif
+
```
Comment (drashna) [inline quantum/encoder.c:180]: This would need to be pulled out and handled as a separate PR.

File: keyboards/keychron/q2/rev_0111/rules.mk [hunk 2]
```diff
@@ -20,7 +20,8 @@ ENCODER_ENABLE = yes        # Enable Encoder
 DIP_SWITCH_ENABLE = yes
 RGB_MATRIX_ENABLE = yes
 RGB_MATRIX_DRIVER = CKLED2001
-EEPROM_DRIVER = i2c
+# EEPROM_DRIVER = i2c
```
Comment (drashna) [inline keyboards/keychron/q2/rev_0111/rules.mk:23]: Why this change?  

This would remove any eeprom support from the board and make via support useless.

**Review by drashna — CHANGES_REQUESTED** [review]

**Comment by tzarc** [comment]
We'll be moving towards a decoupled model with `ENCODER_DRIVER = xxxx` instead of continuing here.

**Pull request closed without merging**